desc_source.go

```diff
@@ -302,3 +294,75 @@ func addFilesToSet(allFiles []*descriptorpb.FileDescriptorProto, expanded map[st
 	}
 	return append(allFiles, fd.AsFileDescriptorProto())
 }
+
+// WriteProtoFiles will use the given descriptor source to resolve all the given
+// symbols and write proto files with their definitions to the given output directory.
+func WriteProtoFiles(outProtoDirPath string, descSource DescriptorSource, symbols ...string) error {
+	filenames, fds, err := getFileDescriptors(symbols, descSource)
+	if err != nil {
+		return err
+	}
+	// now expand that to include transitive dependencies in topologically sorted
+	// order (such that file always appears after its dependencies)
+	expandedFiles := make(map[string]struct{}, len(fds))
+	allFilesSlice := make([]*desc.FileDescriptor, 0, len(fds))
+	for _, filename := range filenames {
+		allFilesSlice = addFilesToFileDescriptorList(allFilesSlice, expandedFiles, fds[filename])
+	}
+	pr := protoprint.Printer{}
+	// now we can serialize to files
+	for _, fd := range allFilesSlice {
+		fdFQName := fd.GetFullyQualifiedName()
+		dirPath := filepath.Dir(fdFQName)
+		outFilepath := filepath.Join(outProtoDirPath, dirPath)
+		if err := os.MkdirAll(outFilepath, 0755); err != nil {
+			return fmt.Errorf("failed to create directory %q: %v", outFilepath, err)
+		}
+		fileName := filepath.Base(fdFQName)
+		filePath := filepath.Join(outFilepath, fileName)
+		f, err := os.Create(filePath)
+		if err != nil {
+			if f != nil {
+				_ = f.Close()
+			}
```

Comment:
you should be able to omit this block, if err is non-nil, f will be nil

Reply:
solved